### ISSUE TITLE
Request: Support Maven releases

### ISSUE THREAD

**Issue by jfairley** [semantic-release/git#123]
[![maven logo](https://maven.apache.org/images/maven-logo-black-on-white.png)](https://maven.apache.org/)

### _TL;DR_

Maven's release cycle includes two commits, the first of which is tagged. I would like to expand the capabilities of `@semantic-release/git` to support the Maven release flow.

### What is Maven?

Maven is the de-facto build and dependency management tool for Java. The configuration file for Maven is the `pom.xml` file, which typically can be found at the root of the project. Not too different from NPM's `package.json` file, the `pom.xml` file includes versioning, dependencies, build and test configurations, and distribution information.

### How does Maven version?

Unlike NPM libraries, a Maven repository at rest represents itself with a "snapshot" version number like _1.0.0-SNAPSHOT_. Building a maven project is almost always synonymous with installing a "snapshot" build into the local repository of dependencies. When versioned with "snapshot" numbering, a Maven repository is a dev and testing phase, which is almost always the case.

Released Maven artifacts are versioned how you would expect based on semantic versioning like _1.0.0_. An artifact versioned like this represents one that is "released" or in "general availility" or any other way of saying that it's ready for widespread use.

Maven knows the difference between _1.0.0-SNAPSHOT_ and _1.0.0_. Specifically, it knows that the former always precedes the latter.

### What is a Maven release?

When an artifact is released using Maven, the versioning goes from "snapshot" to semver to the next "snapshot".

In practice you might expect an artifact's version to change like follows:

1. pre-release:  `<version>1.0.0-SNAPSHOT</version>`
2. released:     `<version>1.0.0</version>`
3. post-release: `<version>1.0.1-SNAPSHOT</version>`

### How does a Maven release affect SCM?

In SCM, it is standard for the tagged commit to reference to the point at which the version represented the release (_1.0.0_).

Therefore, minimally a repository has three commits for a single release: 

- one commit before the tag
	- This is the last commit before starting the release script.
	- current version: _1.0.0-SNAPSHOT_
- one commit which is tagged
	- This commit represents the state of the code during artifact deploy.
	- The release script creates this commit prior to tagging.
	- current version: _1.0.0_
- one commit after the tag
	- This commit prepares the repository for continued development.
	- The release script creates this commit after tagging.
	- current version: _1.0.1-SNAPSHOT_

### How does this relate to `@semantic-release/git`?

Dealing with Maven's release cycle means ideally, two commits need to be made during the overall release, and minimally, any `git push` execution occurs only _after_ these commits exist.

More specifically:

1. The _1.0.0_ change and subsequent commit should be created during the `prepare` step, before the tag and before code is published.
2. While in this _1.0.0_ state, the created commit can be tagged.
3. Still in the _1.0.0_ state, the code should be built and published. This makes the most sense to occur during the `publish` step.
4. To complete the release, the code must be prepared for future work by incrementing back to a "snapshot" number, a change which needs to be committed and pushed to SCM. IMO, this is fine to occur either during `publish` or `success` step, but I expect `semantic-release` to consider this relevant to the `publish` step.

What I would like to see from `@semantic-release/git` is flexibility to create a commit during `prepare` _and_ during `publish`. This does not have to be the default functionality, but allowing for it makes it possible to release with Maven without resorting to `@semantic-release/exec` for SCM actions.

My proposal is available at https://github.com/semantic-release/git/pull/122.

### Further Notes

- It should be noted that the Maven "snapshot" versioning and release versioning is more than just practice. Maven has built-in rule-sets and assumptions about how to deal with artifacts versioned like _1.0.0-SNAPSHOT_ versus _1.0.0_.
- While I'm personally not against the arguments made by `semantic-release` for versioning at-rest artifacts like `0.0.0-development` – it is exactly how I treat my _NPM_ artifacts – it just doesn't fit with Maven.

**Comment by jfairley** [comment]
@pvdlg Any opinions on this?

**Comment by jfairley** [comment]
@pvdlg Have you taken a minute to consider this? The solution is already built and tested in https://github.com/semantic-release/git/pull/122. At this point, it's out of date, but I'll do the work to update it.

**Comment by pvdlg** [comment]
Thanks for the details issue and sorry to have dropped the ball for a while...
I'm trying to wrap my head around that  and what you describe seems to really deviate from what semantic-release is doing, especially updating the version in the `pom.xml` after a publish.

The core concepts of semantic-release are:
- A commit that ends up on the release branch is releasable and as such will be released right away
- The version are determined by semantic-release based on the commit differences since the last version released

Updating the `pom.xml` after a publish goes against that because:
- There is no point in updating the `pom.xml` with a snapshot version for continued development because subsequent commits will be handled by semantic-release and trigger a release if necessary. There is no notion of "continued development" anymore on a release branch. The release branch become a mirror of what's on the Maven registry.
- Updating the `pom.xml` with a snapshot version (with `1.0.1-SNAPSHOT` in you example) implies to "guess" what the next version would be. That next version might be `1.0.1` or `1.1.0` or `2.0.0`.

I used Maven a long time ago so I'm a bit fuzzy on what is actually required and what is considered a common practice. If I remember correctly updating the `pom.xml` to a snapshot version is not required by Maven.
If that's indeed the case I would recommend to use semantic-release as it was intended: publish a release immediately for any relevant commit (`fix`, `feat` and `BREAKING CHANGE` by default).
That mean, just don't do this additional commit to change the version to snapshot for "continued development" as this practice is now irrelevant in an environment where releases are automated and done for each new commit.
Basically you don't need to have this snapshot version in your `pom.xml` to account for the commits on `master` waiting to be released as those won't exist anymore (they will be released right away).

If you don't want to release the commits that end up on `master` right away then I would recommend to look for a different tool than semantic-release. You can probably modify semantic-release or some plugins to fit your needs but that  would never work really well as it goes against the core principle the tools are build around.

If you want to release beta versions that would be distributed to a limited subset of users for test purpose, I would recommend to try semantic-release beta (v16.0.0) which allow you to maintain multiple branches and make releases on different distribution channels.
With v16.0.0 you could release regular versions from `master` and have a `beta`  branch from which you would release snapshot versions. Each new commit on `beta` would release a new snapshot, with a version determined by semantic-release, and when you merge `snapshot` to `master` the snapshots version merged would be made available as regular versions.
You would be able to do "continued development" on the `beta` branch and release code for a subset of users, and make regular releases once you merge `beta` into `master`. In that scenario you wouldn't need to make any post-publish commits to update the `pom.xml`.

In any case what you propose deviate too much from what semantic-release aim to do and we won't implement it in the default plugins. If the solution I proposed doesn't work, you can still develop you own plugins. However I can't guarantee you won't encounter issues in the future, as this won't be a scenario we will actively try to support.

**Comment by jfairley** [comment]
Well, thanks for trying to wrap your head around it. I'll admit that Java artifact management does exactly mimic NPM's, and as an active Node dev, I understand your arguments.

Unfortunately there's just an element of: 

1. Java just does things differently. The snapshot concept, for instance, doesn't even exist in NPM. It's a first-class citizen in Maven and Gradle.
2. Java has done things the way it's done things for a long time. I know from talking to my colleagues during this effort that bring seasoned Java developers into this style of artifact management is practically just setting it up to not be used.

Anyway, I've resolved that this just isn't meant to be. I continue to use `semantic-release` in all of my Node projects, and a co-worker of mine has built an internally used Maven plugin that mimics `semantic-release`. Maybe we'll have time to refine and publicize that, but in the meantime, teams have their processes and work is getting done.

Thanks again. If I ever have any contributions that are Node-related, I'll be sure to submit them.

In the meantime, I've had my eye on v16 beta for quite some time. I look forward to using it, but don't see myself having time to be a beta user. Godspeed, and I look forward to the finished product.
